Extension table

sgraband opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->
Adds a script that logs a csv table with an overview of the impact an extension has on Theia's startup time. The script compares the startup time of a defined base application `scripts/performance/base-package.json` with the base application + an additional extension.
The performance measurement script from #9777 is used to measure the startup time. Therefore this PR depends on that PR. The script can receive multiple extensions, which are then tested one by one. If no extensions are provided all extensions in the `packages` folder will be tested.

Via parameters you can set a 
- baseTime for a previous measurement of the base application
- list of extensions that shall be tested
- number of runs for each measurement
- For more information see the README in `scripts/performance`

Example Result:

|Extension name|MEAN (10 runs) (in s)|STDEV (in ms)|CV (%)|DIFF (in ms)|
|---|---|---|---|---|
|Base theia|2.027|0.084|4.144|-|
|"@theia/git":"1.17.0"|2.103|0.041|1.950|0.076|

Resolves #10191.

Contributed on behalf of STMicroelectronics

Signed-off-by: Simon Graband <sgraband@eclipsesource.com>
#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Run `node ./extension-impact.js` from the `scripts/performance` directory. And have a look at the resulting table. 
Note: This may take a while. To reduce the time either set the runs to a smaller number than 10 or use only a set of extensions.
#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

colin-grant-work reviewed (commented):
Re: the experience, a bit more logging might help. At the moment, the user runs the command, a single line is printed, and then nothing happens for a while. 
Maybe something like this:
```
Measuring base Theia....
Starting run 1...
Run 1 complete in (time)...
Starting run 2...

Measuring (plugin)...
```
The format for specifying the plugins to test is pretty counterintuitive / prone to mistake - maybe just use something like `'@theia/keymaps:1.17.0'` rather than the multiple double quotes - unless that format is likely to produce incorrect parsing?

colin-grant-work commented:
One other problem I noticed is that when I switched to other work, Theia opened at the demo workspace from the performance measurement. It might be nice to reset the recent workspaces JSON file to the state before the performance measurement was started.

sgraband commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,178 @@
+}
+
+async function calculateExtension(extensionQualifier) {
+    let package = JSON.parse(JSON.stringify(basePackage));
```
> Why stringify it and then parse it again?

This is done to create a deep copy of the basePackage Object, as we do not want to modify it, which happens when only calling `const package = basePackage`.

sgraband commented on scripts/performance/extension-impact.js:
```diff
@@ -0,0 +1,178 @@
+}
+
+(async () => {
+    process.on('SIGINT', exitHandler.bind());
+    process.on('exit', exitHandler.bind());
```
You are right. This can be ommited.

sgraband commented:
Thanks for the review @colin-grant-work! I worked in all of your inline comments.

Just a general note: The contents of the first commit (e.g. the `measure-performance.js` file) are part of [this](https://github.com/eclipse-theia/theia/pull/9777) open PR. Could you have another look at it and see if we can merge it? This would ease the review of this PR and any future rebasing ;)

> The format for specifying the plugins to test is pretty counterintuitive / prone to mistake - maybe just use something like '@theia/keymaps:1.17.0' rather than the multiple double quotes - unless that format is likely to produce incorrect parsing?

I tested this again and actually @theia/keymaps:1.17.0 is enough as long as it does not contain any whitespaces. (I updated the readme and the help command)

> Re: the experience, a bit more logging might help. At the moment, the user runs the command, a single line is printed, and then nothing happens for a while.
Maybe something like this:

While i understand where you are coming from, the goal of this script is to get an extension table in csv format, so that you can easily get an overview (either in the command line or in, for example, Excel. When we would log additional messages, this would not work anymore and you would also not be able to save the output of the script into a csv file, which can then be opened.

Maybe we should add a note to the Readme, that this may take a while, and no output is expected?

> One other problem I noticed is that when I switched to other work, Theia opened at the demo workspace from the performance measurement. It might be nice to reset the recent workspaces JSON file to the state before the performance measurement was started.

Is there an easy way to platform independently get the location of that file? If yes, then there should be no problem in adding this. Even though this might also be useful, as the user might want to debug the workspace he opened in the extension table, if any problems occur. WDYT?

colin-grant-work commented:
> Just a general note: The contents of the first commit (e.g. the `measure-performance.js` file) are part of [this](https://github.com/eclipse-theia/theia/pull/9777) open PR. Could you have another look at it and see if we can merge it? This would ease the review of this PR and any future rebasing ;)

That PR looks mostly ready to me, though it looks like some work is necessary to get the builds green - maybe just committing any changes to the `yarn.lock` file.

> While i understand where you are coming from, the goal of this script is to get an extension table in csv format, so that you can easily get an overview (either in the command line or in, for example, Excel. When we would log additional messages, this would not work anymore and you would also not be able to save the output of the script into a csv file, which can then be opened.

I don't think that this objective presents any insurmountable hurdle to providing a better user experience. For example, you could set up a logging utility that sent report-relevant messages to a CSV file at a pre-defined or user-supplied path as well as the console (something like `csvstream.write(message); console.log(message);`), while strictly user-facing messages went only to the console. You could then repeat the report messages in the console at the end of the run so that a summary was available at a glance, as well as as in the report file.

> Is there an easy way to platform independently get the location of that file? If yes, then there should be no problem in adding this. Even though this might also be useful, as the user might want to debug the workspace he opened in the extension table, if any problems occur. WDYT?

There is some complexity here. One solution might be to run Theia with the `THEIA_CONFIG_DIR` environment variable set to some predefined location so that the user's real configuration files aren't affected at all. Then the user could still debug the workspace if desired. But having to switch workspaces isn't such a pain that it needs too much care to avoid.

sgraband commented:
> I don't think that this objective presents any insurmountable hurdle to providing a better user experience. For example, you could set up a logging utility that sent report-relevant messages to a CSV file at a pre-defined or user-supplied path as well as the console (something like csvstream.write(message); console.log(message);), while strictly user-facing messages went only to the console. You could then repeat the report messages in the console at the end of the run so that a summary was available at a glance, as well as as in the report file.

Thats a great suggestion. I added this. The script will create a csv file (location can be passed via parameter) and prints this file at the end of the script. During the runtime the script will give updates about the current process. The output will look like this:
```
Building the browser example with @theia/git:1.18.0.
Measuring the startup time with @theia/git:1.18.0 2 times. This may take a while.
[Performance][Startup][1] Largest Contentful Paint (LCP): 2.170 seconds
[Performance][Startup][2] Largest Contentful Paint (LCP): 2.254 seconds
[Performance][Startup][MEAN] Largest Contentful Paint (LCP): 2.212 seconds
[Performance][Startup][STDEV] Largest Contentful Paint (LCP): 0.042 seconds
@theia/git:1.18.0, 2.212, 0.042, 1.899, 0.212
```

> There is some complexity here. One solution might be to run Theia with the THEIA_CONFIG_DIR environment variable set to some predefined location so that the user's real configuration files aren't affected at all. Then the user could still debug the workspace if desired. But having to switch workspaces isn't such a pain that it needs too much care to avoid.

I added this as well. In the beginning of the script a new, empty, `theia-config-dir` directory is created and passed as an environment variable for the `theia start` command. When the workspace is cleaned, the directory is deleted as well. This way the script has no impact on the recent workspace etc. for the user.

sgraband commented:
So can this be merged?
